shared/actions/chat.js

```diff
@@ -0,0 +1,303 @@
+    param: {
+      conversationID: keyToConversationID(action.payload.conversationIDKey),
+      msg: {
+        clientHeader,
+        messageBody: {
+          messageType: CommonMessageType.text,
+          text: {
+            body: action.payload.text,
+          },
+        },
+      },
+    },
+  })
+}
+
+function * _incomingMessage (action: IncomingMessage): SagaGenerator<any, any> {
+  if (action.payload.activity.ActivityType === NotifyChatChatActivityType.incomingMessage) {
+    const incomingMessage: ?IncomingMessageRPCType = action.payload.activity.IncomingMessage
+    if (incomingMessage) {
+      const messageUnboxed: MessageUnboxed = incomingMessage.message
+      const yourNameSelector = (state: TypedState) => state.config.username
+      const yourName = yield select(yourNameSelector)
+      const message = _threadToStorable(messageUnboxed, 0, yourName)
+
+      yield put({
+        type: Constants.appendMessages,
+        payload: {
+          conversationIDKey: conversationIDToKey(incomingMessage.convID),
+          messages: [message],
+        },
+      })
+    }
+  }
+}
+
+function * _setupNewChatHandler (): SagaGenerator<any, any> {
+  yield put((dispatch: Dispatch) => {
+    engine().setIncomingHandler('chat.1.NotifyChat.NewChatActivity', ({uid, activity}) => {
+      dispatch({type: Constants.incomingMessage, payload: {activity}})
+    })
+  })
+}
+
+function * _loadInbox (): SagaGenerator<any, any> {
+  // $ForceType
+  const inbox: GetInboxAndUnboxLocalRes = yield call(localGetInboxAndUnboxLocalRpcPromise, {params: {}})
+  const conversations: List<InboxState> = _inboxToConversations(inbox)
+  yield put({type: Constants.loadedInbox, payload: {inbox: conversations}})
+}
+
+function * _loadedInbox (action: LoadedInbox): SagaGenerator<any, any> {
+  const selector = (state: TypedState) => state.chat.get('selectedConversation')
+  const selectedConversation = yield select(selector)
+
+  if (!selectedConversation) {
+    if (action.payload.inbox.count()) {
+      const mostRecentConversation = action.payload.inbox.get(0)
+      yield put(selectConversation(mostRecentConversation.get('conversationIDKey')))
+    }
+  }
+}
+
+function * _loadMoreMessages (): SagaGenerator<any, any> {
+  const selectedSelector = (state: TypedState) => state.chat.get('selectedConversation')
+  const conversationIDKey = yield select(selectedSelector)
+
+  if (!conversationIDKey) {
+    return
+  }
+
+  const conversationID = keyToConversationID(conversationIDKey)
+  const conversationStateSelector = (state: TypedState) => state.chat.get('conversationStates', Map()).get(conversationIDKey)
+  const oldConversationState = yield select(conversationStateSelector)
+
+  let next
+  if (oldConversationState) {
+    if (oldConversationState.get('isLoading')) {
+      __DEV__ && console.log('Bailing on chat load more due to isloading already')
+      return
+    }
+
+    if (!oldConversationState.get('moreToLoad')) {
```

Comment:
This is getting older messages. We can't be confused about this state. Newer messages come through gregor, not this.